Userland/xargs.cpp

```diff
@@ -0,0 +1,304 @@
+
+    if (is_stdin) {
+        devnull_fd = open("/dev/null", O_RDONLY | O_CLOEXEC);
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    size_t total_command_length = 0;
+    size_t items_used_for_this_command = 0;
+
+    auto fail = read_items(fp, entry_separator, [&](StringView item) {
+        if (item.ends_with('\n'))
+            item = item.substring_view(0, item.length() - 1);
+
+        if (item.is_empty())
+            return Continue;
+
+        // The first item is processed differently, as all the initial-arguments are processed _with_ that item
+        // as their substitution target (assuming that substitution is enabled).
+        // Note that if substitution is not enabled, we manually insert a substitution target at the end of initial-arguments,
+        // so this item has a place to go.
+        if (items_used_for_this_command == 0) {
+            child_argv.ensure_capacity(initial_arguments.size());
+
+            initial_arguments.for_each_joined_argument(item, [&](const String& string) {
+                total_command_length += string.length();
+                child_argv.append(strdup(string.characters()));
+            });
+
+            ++items_used_for_this_command;
+        } else {
+            if ((max_lines > 0 && items_used_for_this_command + 1 > max_lines) || total_command_length + item.length() + 1 >= max_bytes) {
+                if (!run_command(child_argv, verbose, is_stdin, devnull_fd))
+                    return Stop;
+                items_used_for_this_command = 0;
+                total_command_length = 0;
+                return Unget;
+            } else {
+                child_argv.append(strndup(item.characters_without_null_termination(), item.length()));
+                ++items_used_for_this_command;
+            }
+        }
+
+        return Continue;
+    });
+
+    if (!fail && !child_argv.is_empty())
+        fail = !run_command(child_argv, verbose, is_stdin, devnull_fd);
+
+    if (!is_stdin)
+        fclose(fp);
+
+    return fail ? 1 : 0;
+}
+
+bool read_items(FILE* fp, char entry_separator, Function<Decision(StringView)> callback)
+{
+    bool fail = false;
+
+    for (;;) {
+        char* item = nullptr;
+        size_t buffer_size = 0;
+
+        auto item_size = getdelim(&item, &buffer_size, entry_separator, fp);
+
+        if (item_size < 0) {
+            // getdelim() will return -1 and set errno to 0 on EOF.
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        Decision decision;
+        do {
+            decision = callback(item);
+            if (decision == Stop) {
+                free(item);
+                return true;
+            }
+        } while (decision == Unget);
+
+        free(item);
+    }
+
+    return fail;
+}
+
+bool run_command(Vector<char*>& child_argv, bool verbose, bool is_stdin, int devnull_fd)
```

Comment:
Actually let's make this

```suggestion
bool run_command(Vector<char*>&& child_argv, bool verbose, bool is_stdin, int devnull_fd)
```

and change the call sites to do an explicit `move()`.

Reply:
I'm gonna have to ask for a reason, as moving this is entirely too wrong to me, we're not giving the vector to `run_command`, we're giving the contents to it.

i.e. moving it out will force us to recreate a new Vector for each iteration, while we can reuse the same backing memory and not reallocate a billion vectors

Reply:
I think you might misunderstand what `&&` and `move()` do in C++. They don't actually move anything; they just signify to the reader (and to the compiler, when resolving overloads, see below) that you do not care about this value staying the same (or valid) after the call.

Now, if the method you're invoking is overloaded and has separate versions that take a `const T&` and a `T&&`, the former would usually copy the `T` value, and the latter would "steal" the value from the caller, as the caller indicated it doesn't care about the value.

Here, `run_command` empties (and otherwise cleans up) the vector, which to me sounds exactly like what a function taking a `&&` can be expected to do. It should have *zero* impact on the generated code (feel free to verify), 'cause we're not overloading anything, but it serves as a sort of comment for the reader.

Reply:
to my information, there is never any copy involved with passing a reference, rvalue or lvalue.
you might end up with a copy constructor, but that's completely irrelevant here.

so far as the compiler (and the reader) is concerned, the value left behind after an rvalue reference is passed to some function is completely crap data.
Even though I'm pretty sure that will not mess with the value of our Vector here, it sends a completely wrong message.

I _do_ care about the vector instance, if our Vector impl someday decided to actually move on a move constructor from an rvalue reference, this code will be completely screwed.

Reply:
Update: our Vector instance _does_ actually move on `Vector(Vector&&)`, so I'm gonna go ahead and say that this _would_ actually reallocate the vector every time

Reply:
> if our Vector impl someday decided to actually move on a move constructor from an rvalue reference, this code will be completely screwed.

I bet it already does. But we're not move-constructing any vectors here, just passing one rvalue Vector reference to a function.

Reply:
Why fight on theoretical stuff when we can fight over profiles?
let's see if there actually is a lot of reallocations :shrug:

Reply:
Fantastic.
The generated code is literally the same, equal offsets, equal instructions.
Same number of everything.

Reply:
What did I tell ya? :wink:

> It should have zero impact on the generated code (feel free to verify), 'cause we're not overloading anything, but it serves as a sort of comment for the reader.

Reply:
At this point, even though I find the notion of telling the compiler to `move` the vector very weird for the use-case...I don't mind "move"'ing the vector.

I will say that doing this might confuse future readers, however.

Reply:
Also, `clear_with_capacity` is so very misleading, it reads to me as "clear everything, including the capacity"

Reply:
> Also, `clear_with_capacity` is so very misleading, it reads to me as "clear everything, including the capacity"

Agreed, perhaps rename it to `clear_keeping_capacity()` (or `clear_preserving_capacity()`) if you're messing with AK anyway (and if @awesomekling approves)?

Reply:
I went ahead and added a comment about this move behaviour.
commenting a comment ftw